hbase-server/src/main/java/org/apache/hadoop/hbase/master/webapp/MetaBrowser.java

```diff
@@ -0,0 +1,378 @@
+    if (requestValueStr == null) {
+      return null;
+    }
+    final RegionState.State requestValue = tryValueOf(RegionState.State.class, requestValueStr);
+    if (requestValue == null) {
+      errorMessages.add(buildScanRegionStateMalformedErrorMessage(requestValueStr));
+      return null;
+    }
+    return requestValue;
+  }
+
+  private static byte[] resolveScanStart(final HttpServletRequest request) {
+    final String requestValue = resolveRequestParameter(request, SCAN_START_PARAM);
+    if (requestValue == null) {
+      return null;
+    }
+    return Bytes.toBytesBinary(requestValue);
+  }
+
+  private static TableName resolveScanTable(final HttpServletRequest request) {
+    final String requestValue = resolveRequestParameter(request, SCAN_TABLE_PARAM);
+    if (requestValue == null) {
+      return null;
+    }
+    return TableName.valueOf(requestValue);
+  }
+
+  private static String resolveRequestParameter(final HttpServletRequest request,
+    final String param) {
+    if (request == null) {
+      return null;
+    }
+    final String requestValueStrEnc = request.getParameter(param);
+    if (StringUtils.isBlank(requestValueStrEnc)) {
+      return null;
+    }
+    return urlDecode(requestValueStrEnc);
+  }
+
+  private static Filter buildTableFilter(final TableName tableName) {
+    return new PrefixFilter(tableName.toBytes());
+  }
+
+  private static Filter buildScanRegionStateFilter(final RegionState.State state) {
+    return new SingleColumnValueFilter(
+      HConstants.CATALOG_FAMILY,
+      HConstants.STATE_QUALIFIER,
+      CompareOperator.EQUAL,
+      // use the same serialization strategy as found in MetaTableAccessor#addRegionStateToPut
+      Bytes.toBytes(state.name()));
+  }
+
+  private Filter buildScanFilter() {
+    if (scanTable == null && scanRegionState == null) {
+      return null;
+    }
+
+    final List<Filter> filters = new ArrayList<>(2);
+    if (scanTable != null) {
+      filters.add(buildTableFilter(scanTable));
+    }
+    if (scanRegionState != null) {
+      filters.add(buildScanRegionStateFilter(scanRegionState));
+    }
+    if (filters.size() == 1) {
+      return filters.get(0);
+    }
+    return new FilterList(FilterList.Operator.MUST_PASS_ALL, filters);
+  }
+
+  private Scan buildScan() {
+    final Scan metaScan = new Scan()
+      .addFamily(HConstants.CATALOG_FAMILY)
+      .readVersions(1)
+      .setLimit((scanLimit != null ? scanLimit : SCAN_LIMIT_DEFAULT) + 1);
+    if (scanStart != null) {
+      metaScan.withStartRow(scanStart, false);
+    }
+    final Filter filter = buildScanFilter();
+    if (filter != null) {
+      metaScan.setFilter(filter);
+    }
+    return metaScan;
+  }
+
+  /**
+   * Adds {@code value} to {@code encoder} under {@code paramName} when {@code value} is non-null.
+   */
+  private <T> void addParam(final QueryStringEncoder encoder, final String paramName,
```

Comment:
nit: If toString() is the only needed call, no need to use generics / templatize it? Just pass an object?

Reply:
Meh, I suppose.